queasars/minimum_eigensolvers/evqe/evolutionary_algorithm/selection.py

```diff
@@ -0,0 +1,110 @@
+                [individual.get_parameter_values()],
+            )
+            for individual in population.individuals
+        ]
+
+        wait(future_evaluation_results)
+        evaluation_results: list[float] = [future.result()[0] for future in future_evaluation_results]
+        del future_evaluation_results
+        operator_context.circuit_evaluation_count_callback(len(population.individuals))
+
+        # cannot apply selection if speciation was not done beforehand
+        if (
+            population.species_representatives is None
+            or population.species_members is None
+            or population.species_membership is None
+        ):
+            raise EVQESelectionException(
+                "Selection can't be finished if speciation information is missing!\n"
+                + "Either the species_representatives, species_members or species_membership\n"
+                + "attribute of the population is None!"
+            )
+
+        # report the result of the evaluation
+        best_individual_index: int = int(argmin(evaluation_results))
+        result: BasePopulationEvaluationResult[EVQEIndividual] = BasePopulationEvaluationResult(
+            population=population,
+            expectation_values=tuple(evaluation_results),
+            best_individual=population.individuals[best_individual_index],
+            best_expectation_value=evaluation_results[best_individual_index],
+        )
+        operator_context.result_callback(result)
+
+        # disallow negative or 0 values in fitnesses by shifting all evaluation results by a fixed offset
+        offset: float
+        if evaluation_results[best_individual_index] <= 0:
+            offset = -evaluation_results[best_individual_index] + 1
+        else:
+            offset = 0
+
+        fitness_values: list[float] = [
+            (
+                evaluation_results[i]
+                + offset
+                + self.alpha_penalty * len(individual.layers)
+                + self.beta_penalty * individual.get_n_controlled_gates()
+            )
+            * float(len(population.species_members[population.species_membership[i]]))
+            for i, individual in enumerate(population.individuals)
+        ]
+
+        fitness_weights: list[float] = [1 / fitness for fitness in fitness_values]
+
+        selected_individuals: list[EVQEIndividual] = self.random_generator.choices(
+            population.individuals, weights=fitness_weights, k=len(population.individuals)
+        )
+
+        return EVQEPopulation(
+            individuals=tuple(selected_individuals),
+            species_representatives=population.species_representatives,
+            species_members=None,
+            species_membership=None,
+        )
+
+    def get_n_expected_circuit_evaluations(
+        self, population: EVQEPopulation, operator_context: OperatorContext
+    ) -> Optional[int]:
+        return len(population.individuals)
```

Comment:
Ah I think we discussed this already or I missed it above: With "circuit evaluations" you mean the count without the factor of shots in case of a sampler? Because this formula here means every single individual counts once.

Reply:
Yes this only counts how often the sampler / estimator is invoked to evaluate a quantum circuit / parameter combination.  This is intended, as the shots are set by the user as an option in the Sampler / Estimator and are inaccessible to this algoritm.

Reply:
Ah, understood, thanks!